### PR TITLE
Feature/remove data after eject reducer

### PR DESCRIPTION
**Pull request by mmaarrttyy**

### CODE REVIEW

File: packages/epics/package.json [hunk 1]
```diff
@@ -20,12 +20,12 @@
     "ramda-extension": "^0.9.1"
   },
   "peerDependencies": {
-    "redux": "^4.0.1",
+    "redux": "4.0.4",
```
Comment (wafflepie) [inline packages/epics/package.json]: leave the caret here in peerDependencies, please

File: packages/middleware/package.json [hunk 2]
```diff
@@ -20,7 +20,7 @@
     "ramda-extension": "^0.9.1"
   },
   "peerDependencies": {
-    "redux": "^4.0.1"
```
Comment (wafflepie) [inline packages/middleware/package.json]: leave the caret here in peerDependencies, please

File: packages/reducers/package.json [hunk 3]
```diff
@@ -20,7 +20,7 @@
     "ramda-extension": "^0.9.1"
   },
   "peerDependencies": {
-    "redux": "^4.0.1"
```
Comment (wafflepie) [inline packages/reducers/package.json]: leave the caret here in peerDependencies, please

File: packages/reducers/src/makeEnhancer.test.js [hunk 4]
```diff
@@ -44,7 +46,18 @@ describe('makeEnhancer', () => {
 		expect(getEntries(store)).toEqual([]);
 	});
 
-	it('throws when injecting a function without a namespace', () => {
```
Comment (wafflepie) [inline packages/reducers/src/makeEnhancer.test.js]: why did you remove this test? :)

File: packages/reducers/src/makeEnhancer.test.js [hunk 5]
```diff
@@ -44,7 +46,18 @@ describe('makeEnhancer', () => {
...
+		const reducerA = (state = { name: 'a' }) => state;
+
+		store.injectReducers({ a: reducerA }, { namespace: 'nsA' });
+		expect(store.getState()).toEqual({
+			namespaces: {
+				nsA: { a: { name: 'a' } },
+			},
+		});
+
+		store.ejectReducers(identity, { namespace: 'nsB' });
```
Comment (wafflepie) [inline packages/reducers/src/makeEnhancer.test.js]: nooo, this should not be the expected behaviour.

the state clean-up should only happen for the appropriate reducer

	store.ejectReducers(identity, { namespace: 'nsB' });

this code should not do anything to our state, because identity was never present in the store!

what we need to do is check that

```store.ejectReducers({ a: reducerA }, { namespace: 'nsA' });```

doesnt leave any traces of the ejected reducer's state

Comment (wafflepie) [reply packages/reducers/src/makeEnhancer.test.js]: change this test and please add another test to ensure that the state of other reducers is left untouched

File: packages/reducers/src/makeEnhancer.test.js [hunk 6]
```diff
@@ -1,11 +1,13 @@
 import { identity } from 'ramda';
 import { FUNCTION_KEY } from '@redux-tools/injectors';
 import { DEFAULT_FEATURE } from '@redux-tools/namespaces';
+import { createStore as createStoreRedux } from 'redux';
 
 import makeEnhancer, { storeInterface } from './makeEnhancer';
 
 const createStore = () => ({
 	replaceReducer: jest.fn(),
+	state: {},
```
Comment (wafflepie) [inline packages/reducers/src/makeEnhancer.test.js]: i dont think this line is necessary

besides, a store doesnt have state, but a getState method

File: packages/reducers/src/combineReducers.test.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
+import { createStore } from 'redux';
+import combineReducers from './combineReducers';
+
+describe('Utils', () => {
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: remove this Utils wrapper

File: packages/reducers/src/combineReducers.test.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+			// we double-cast because these conditions can only happen in a javascript setting
+			const reducer = combineReducers({
+				fake: true,
+				broken: 'string',
+				another: { nested: 'object' },
+				stack: (state = []) => state,
+			});
+			expect(Object.keys(reducer(undefined, { type: 'push' }))).toEqual(['stack']);
+		});
+		it('allows a symbol to be used as an action type', () => {
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: remove this test pls

File: packages/reducers/src/combineReducers.test.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+			const reducer = combineReducers({
+				counter: (state = 0, action) => (action.type === 'increment' ? state + 1 : state),
+				stack: (state = [], action) => (action.type === 'push' ? [...state, action.value] : state),
+			});
+			const s1 = reducer(undefined, { type: 'increment' });
+			expect(s1).toEqual({ counter: 1, stack: [] });
+			const s2 = reducer(s1, { type: 'push', value: 'a' });
+			expect(s2).toEqual({ counter: 1, stack: ['a'] });
+		});
+		it('ignores all props which are not a function', () => {
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: id prefer if you used newlines inbetween the `it` calls :)

File: packages/reducers/src/combineReducers.test.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+						case increment:
+							return state + 1;
+						default:
+							return state;
+					}
+				},
+			});
+			expect(reducer({ counter: 0 }, { type: increment }).counter).toEqual(1);
+		});
+		it('maintains referential equality if the reducers it is combining do', () => {
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: this test is really useful!

File: packages/reducers/src/combineReducers.test.js [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+					}
+				},
+				child3(state = {}) {
+					return state;
+				},
+			});
+			const initialState = reducer(undefined, { type: '@@INIT' });
+			expect(reducer(initialState, { type: 'increment' })).not.toBe(initialState);
+		});
+		describe('With Replace Reducers', function() {
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: i think this block of tests is a little misleading, please remove the entire WithReplaceReducers block

File: packages/reducers/src/combineReducers.test.js [hunk 12]
```diff
@@ -0,0 +1,125 @@
+import { createStore } from 'redux';
+import combineReducers from './combineReducers';
+
+describe('Utils', () => {
+	describe('combineReducers', () => {
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: i'm missing a test for the following behaviour:

`it('preserves unknown keys in the state'`

contact me if you don't understand what should be the purpose of this test

File: packages/reducers/src/combineReducers.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
+import { length, mapObjIndexed, pickBy, keys, o } from 'ramda';
+import { isFunction } from 'ramda-extension';
+
+export default reducers => {
+	const finalReducers = pickBy(isFunction, reducers);
+	const finalReducerKeys = keys(finalReducers);
+
+	return function combination(state = {}, action) {
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.js]: just return an arrow fn

File: packages/reducers/src/combineReducers.test.js (hunk 12 above)
Comment (wafflepie) [reply packages/reducers/src/combineReducers.test.js]: i think that it will definitely result in some changes in your combineReducers implementation :) i'm curious in how you're going to tackle that!

File: packages/injectors/src/enhanceStore.js [hunk 14]
```diff
@@ -53,7 +53,10 @@ const enhanceStore = (prevStore, storeInterface, { onEjected = noop, onInjected
 		dispatch({
 			type: `@redux-tools/${actionType}_EJECTED`,
 			payload: keys(injectables),
-			meta: props,
+			meta: {
+				...props,
+				isFunctionReducer: o(length, keys)(injectables) === 0 && length(entries) > 0,
```
Comment (tommmyy) [inline packages/injectors/src/enhanceStore.js]: o(length, keys) can be its own util function

File: packages/reducers/src/combineReducers.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import { length, mapObjIndexed, pickBy, keys, o } from 'ramda';
+import { isFunction } from 'ramda-extension';
+
+export default reducers => {
+	const finalReducers = pickBy(isFunction, reducers);
```
Comment (tommmyy) [inline packages/reducers/src/combineReducers.js]: pickBy(isFunction) can be its own util function

File: packages/reducers/src/combineReducers.js [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+
+		mapObjIndexed(key => {
+			const reducer = finalReducers[key];
+			const previousStateForKey = state[key];
+			const nextStateForKey = reducer(previousStateForKey, action);
+			nextState[key] = nextStateForKey;
+			hasChanged = hasChanged || nextStateForKey !== previousStateForKey;
+		}, finalReducerKeys);
+
+		hasChanged = hasChanged || length(finalReducerKeys) !== o(length, keys)(state);
```
Comment (tommmyy) [inline packages/reducers/src/combineReducers.js]: o(length, keys) - see my other comments

File: packages/reducers/src/makeEnhancer.js [hunk 17]
```diff
@@ -8,6 +23,29 @@ import composeReducers from './composeReducers';
 
 export const storeInterface = makeStoreInterface('reducers');
 
+const cleanUpReducer = (state, action) => {
+	if (action.type !== '@redux-tools/REDUCERS_EJECTED') return state;
```
Comment (tommmyy) [inline packages/reducers/src/makeEnhancer.js]: use curly braces

File: packages/reducers/src/makeEnhancer.js [hunk 18]
```diff
@@ -8,6 +23,29 @@ import composeReducers from './composeReducers';
...
+
+	const namespaceGroup = action.meta.feature || DEFAULT_FEATURE;
```
Comment (tommmyy) [inline packages/reducers/src/makeEnhancer.js]: this will fail if `meta` is not in action

File: packages/reducers/src/makeEnhancer.js [hunk 19]
```diff
@@ -8,6 +23,29 @@ import composeReducers from './composeReducers';
...
+
+	const pathToNamespace = [namespaceGroup, action.meta.namespace];
+	const pathToReducer = when(always(action.meta.namespace), concat(pathToNamespace))([]);
+
+	const cleanReducers = reduce((state, reducerName) =>
+		dissocPath(append(reducerName, pathToReducer), state)
+	)(__, action.payload);
+
+	const cleanEmptyObjects = compose(
+		when(o(isEmpty, path([namespaceGroup])), dissocPath([namespaceGroup])),
```
Comment (tommmyy) [inline packages/reducers/src/makeEnhancer.js]: lens can be used -> u are updating and reading prop from same path

File: packages/reducers/src/makeEnhancer.js [hunk 20]
```diff
@@ -8,6 +23,29 @@ import composeReducers from './composeReducers';
...
+		when(
+			either(o(isEmpty, path(pathToNamespace)), always(action.meta.isFunctionReducer)),
```
Comment (tommmyy) [inline packages/reducers/src/makeEnhancer.js]: action.meta.isFunctionReducer - this will fail if `meta` is not in action

File: packages/reducers/src/combineReducers.test.js [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+			stack: (state = [], action) => (action.type === 'push' ? [...state, action.value] : state),
+		});
+		const s1 = reducer(undefined, { type: 'increment' });
+		expect(s1).toEqual({ counter: 1, stack: [] });
+		const s2 = reducer(s1, { type: 'push', value: 'a' });
+		expect(s2).toEqual({ counter: 1, stack: ['a'] });
+	});
+
+	it('ignores all props which are not a function', () => {
+		// we double-cast because these conditions can only happen in a javascript setting
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: what is this comment even

File: packages/reducers/src/combineReducers.test.js [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+		const store = createStore(originalCompositeReducer);
+		store.dispatch(ACTION);
+		const initialState = store.getState();
+		store.replaceReducer(combineReducers({ baz, bar }));
+		store.dispatch(ACTION);
+		const nextState = store.getState();
+		expect(nextState).not.toBe(initialState);
+	});
+
+	it('should return the same state when reducers passed to combineReducers not changed', function() {
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: are not

File: packages/reducers/src/combineReducers.test.js [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+		store.dispatch(ACTION);
+		const initialState = store.getState();
+		store.replaceReducer(combineReducers({ foo, bar }));
+		store.dispatch(ACTION);
+		const nextState = store.getState();
+		expect(nextState).toBe(initialState);
+	});
+
+	it(
+		'should return an updated state when one of more reducers ' +
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: change this test to `should preserve state when one or more reducers are removed`

Comment (wafflepie) [reply packages/reducers/src/combineReducers.test.js]: and update content accordingly

File: packages/reducers/src/combineReducers.test.js [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+		const reducer = combineReducers({
+			fake: true,
+			broken: 'string',
+			another: { nested: 'object' },
+			stack: (state = []) => state,
+		});
+		expect(Object.keys(reducer(undefined, { type: 'push' }))).toEqual(['stack']);
+	});
+
+	it('does not have referential equality if one of the reducers changes something', () => {
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: did we get rid of the test where we check that we have maintained referential equality?

pls return it, that's important.


```
	it('maintains referential equality if the reducers it is combining do', () => {
			const reducer = combineReducers({
				child1(state = {}) {
					return state;
				},
				child2(state = {}) {
					return state;
				},
				child3(state = {}) {
					return state;
				},
			});
			const initialState = reducer(undefined, { type: '@@INIT' });
			expect(reducer(initialState, { type: 'FOO' })).toBe(initialState);
```

found it

File: packages/reducers/src/combineReducers.test.js [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+			},
+			child3(state = {}) {
+				return state;
+			},
+		});
+		const initialState = reducer(undefined, { type: '@@INIT' });
+		expect(reducer(initialState, { type: 'increment' })).not.toBe(initialState);
+	});
+
+	it('should return an updated state when additional reducers are passed to combineReducers', function() {
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: use arrow functions everywhere pls

Comment (wafflepie) [reply packages/reducers/src/combineReducers.test.js]: just search for the function keyword in the codebase and get rid of it :D yuck

File: packages/reducers/src/makeEnhancer.js (hunk 18 above)
Comment (wafflepie) [reply packages/reducers/src/makeEnhancer.js]: use `feature`, not `namespaceGroup`

File: packages/reducers/src/makeEnhancer.js [hunk 26]
```diff
@@ -8,6 +23,29 @@ import composeReducers from './composeReducers';
 
 export const storeInterface = makeStoreInterface('reducers');
 
+const cleanUpReducer = (state, action) => {
+	if (action.type !== '@redux-tools/REDUCERS_EJECTED') return state;
+
+	const namespaceGroup = action.meta.feature || DEFAULT_FEATURE;
+
+	const pathToNamespace = [namespaceGroup, action.meta.namespace];
+	const pathToReducer = when(always(action.meta.namespace), concat(pathToNamespace))([]);
```
Comment (wafflepie) [inline packages/reducers/src/makeEnhancer.js]: unnecessary complexity, ˙action.meta.namespace ? pathToNamespace : []˙
 
this will also fail if action.meta is not defined, we have util functions for `getNamespaceByAction` and maybe even `getFeatureByAction`, use them!

File: packages/reducers/src/makeEnhancer.js (hunk 20 above)
Comment (wafflepie) [reply packages/reducers/src/makeEnhancer.js]: isFunctionReducer will hopefully be removed anyway (just a note, see my comment in the other file)

File: packages/reducers/src/makeEnhancer.js [hunk 27]
```diff
@@ -8,6 +23,29 @@ import composeReducers from './composeReducers';
 
 export const storeInterface = makeStoreInterface('reducers');
 
+const cleanUpReducer = (state, action) => {
```
Comment (wafflepie) [inline packages/reducers/src/makeEnhancer.js]: grammar note, I'd personally use `cleanupReducer`

File: packages/reducers/src/makeEnhancer.js [hunk 28]
```diff
@@ -8,6 +23,29 @@ import composeReducers from './composeReducers';
...
+	if (action.type !== '@redux-tools/REDUCERS_EJECTED') return state;
+
+	const namespaceGroup = action.meta.feature || DEFAULT_FEATURE;
+
+	const pathToNamespace = [namespaceGroup, action.meta.namespace];
+	const pathToReducer = when(always(action.meta.namespace), concat(pathToNamespace))([]);
+
+	const cleanReducers = reduce((state, reducerName) =>
```
Comment (wafflepie) [inline packages/reducers/src/makeEnhancer.js]: whew, uhh.. pointfree is not always the way to go, I think that doing sth like

```
const removeEjectedState = prevState => reduce((nextState, reducerKey) => dissoc(), prevState, action.payload))
```

would be more readable (along with the renames)

File: packages/reducers/src/makeEnhancer.js [hunk 29]
```diff
@@ -8,6 +23,29 @@ import composeReducers from './composeReducers';
...
+		dissocPath(append(reducerName, pathToReducer), state)
+	)(__, action.payload);
+
+	const cleanEmptyObjects = compose(
```
Comment (wafflepie) [inline packages/reducers/src/makeEnhancer.js]: prefer `o` for composing 2 unary functions

File: packages/reducers/src/makeEnhancer.js (hunk 19 above)
Comment (wafflepie) [reply packages/reducers/src/makeEnhancer.js]: maybe creating a new function like Lens => PrevState => NextState would be better? :P

File: packages/reducers/src/makeEnhancer.test.js [hunk 30]
```diff
@@ -47,4 +49,135 @@ describe('makeEnhancer', () => {
 	it('throws when injecting a function without a namespace', () => {
 		expect(() => store.injectReducers(identity)).toThrow();
 	});
+
+	it('removes data from state after ejecting (object reducers without preloaded state)', () => {
+		const store = createStoreRedux(state => state, makeEnhancer());
```
Comment (wafflepie) [inline packages/reducers/src/makeEnhancer.test.js]: we already have `let store` in the outer scope :(

File: packages/reducers/src/makeEnhancer.test.js [hunk 31]
```diff
@@ -47,4 +49,135 @@ describe('makeEnhancer', () => {
...
+			nsGroup: {
+				nsA: { a: { name: 'a' } },
+			},
+		});
+
+		store.ejectReducers({ a: reducerA }, { namespace: 'nsA', feature: 'nsGroup' });
+		expect(store.getState()).toEqual({});
+	});
+
+	it('removes data from state after ejecting (object reducers wit preloaded state)', () => {
```
Comment (wafflepie) [inline packages/reducers/src/makeEnhancer.test.js]: with

File: packages/reducers/src/makeEnhancer.test.js [hunk 32]
```diff
@@ -47,4 +49,135 @@ describe('makeEnhancer', () => {
...
+		expect(store.getState()).toEqual({
+			namespaces: {
+				nsA: { a: { name: 'a' } },
+			},
+		});
+
+		store.ejectReducers({ a: reducerA }, { namespace: 'nsA' });
+		expect(store.getState()).toEqual({});
+
+		store.injectReducers({ a: reducerA }, { namespace: 'nsA', feature: 'nsGroup' });
```
Comment (wafflepie) [inline packages/reducers/src/makeEnhancer.test.js]: stop using nsGroup in the codebase pls, more terms are not gonna help anybody

File: packages/reducers/src/combineReducers.test.js [hunk 33]
```diff
@@ -0,0 +1,110 @@
...
+	});
+
+	it('should preserve state when one or more reducers are removed', () => {
+		const originalCompositeReducer = combineReducers({ foo, bar });
+		const store = createStore(originalCompositeReducer);
+		store.dispatch(ACTION);
+		const initialState = store.getState();
+		store.replaceReducer(combineReducers({ bar }));
+		const nextState = store.getState();
+		expect(nextState).not.toBe(initialState);
```
Comment (wafflepie) [inline packages/reducers/src/combineReducers.test.js]: wait wait wait, what are you testing here?

"should preserve state when one or more reducers are removed" does not imply that the state has changed, quite the opposite!

the state should not have changed here (referentially) at all!

please change this test to be `expect(nextState).toBe(initialState)`

i think the solution is to leave out the `nextStateForKey !== previousStateForKey` in `hasChanged = hasChanged || nextStateForKey !== previousStateForKey`

that will make it easier for you to refactor into the `reduce` variant, at least you can practice that :P
